codec-http/src/main/java/io/netty/handler/codec/http/HttpServerCodec.java

```diff
@@ -69,4 +80,62 @@ public HttpServerCodec(int maxInitialLineLength, int maxHeaderSize, int maxChunk
     public void upgradeFrom(ChannelHandlerContext ctx) {
         ctx.pipeline().remove(this);
     }
+
+    private final class HttpServerRequestDecoder extends HttpRequestDecoder {
+        public HttpServerRequestDecoder(int maxInitialLineLength, int maxHeaderSize, int maxChunkSize) {
+            super(maxInitialLineLength, maxHeaderSize, maxChunkSize);
+        }
+
+        public HttpServerRequestDecoder(int maxInitialLineLength, int maxHeaderSize, int maxChunkSize,
+                                        boolean validateHeaders) {
+            super(maxInitialLineLength, maxHeaderSize, maxChunkSize, validateHeaders);
+        }
+
+        public HttpServerRequestDecoder(int maxInitialLineLength, int maxHeaderSize, int maxChunkSize,
+                                        boolean validateHeaders, int initialBufferSize) {
+            super(maxInitialLineLength, maxHeaderSize, maxChunkSize, validateHeaders, initialBufferSize);
+        }
+
+        @Override
+        protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> out) throws Exception {
+            int oldSize = out.size();
+            super.decode(ctx, buffer, out);
+            int size = out.size();
+            for (int i = oldSize; i < size; i++) {
+                Object obj = out.get(i);
+                if (obj instanceof HttpRequest) {
+                    queue.add(((HttpRequest) obj).method());
+                }
+            }
+        }
+    }
+
+    private final class HttpServerResponseEncoder extends HttpResponseEncoder {
+
+        @Override
+        boolean encodeNonChunkedContent(ByteBuf buf, Object msg, long contentLength, List<Object> out) {
+            if (msg instanceof LastHttpContent) {
+                HttpMethod method = queue.poll();
+                if (HttpMethod.HEAD.equals(method)) {
+                    ((LastHttpContent) msg).release();
+                    out.add(buf == null ? Unpooled.EMPTY_BUFFER : buf);
+                    return true;
+                }
+            }
+            return super.encodeNonChunkedContent(buf, msg, contentLength, out);
+        }
+
+        @Override
+        boolean encodeChunkedContent(ChannelHandlerContext ctx, Object msg, long contentLength, List<Object> out) {
+            if (msg instanceof LastHttpContent) {
+                HttpMethod method = queue.poll();
+                if (HttpMethod.HEAD.equals(method)) {
+                    ((LastHttpContent) msg).release();
+                    out.add(Unpooled.EMPTY_BUFFER);
+                    return true;
+                }
```

Comment:
@normanmaurer - I think what @trustin was suggesting was to do something like this:


```java
class HttpObjectEncoder {
  final boolean encodeNonChunkedContent(ByteBuf buf, Object msg, long contentLength, List<Object> out) {
  // use isContentAlwaysEmpty
  }

  final boolean encodeChunkedContent(ChannelHandlerContext ctx, Object msg, long contentLength, List<Object> out) {
  // use isContentAlwaysEmpty
  }

  protected boolean isContentAlwaysEmpty(HttpMessage msg) {
    return false;
  }
}

// In HttpServerCodec.java
private final class HttpServerResponseEncoder extends HttpResponseEncoder {
  @Override
  protected boolean isContentAlwaysEmpty(HttpMessage msg) {
    if (msg instanceof LastHttpContent) {
      HttpMethod method = queue.poll();
      return HttpMethod.HEAD.equals(method);
    }
    return false;
  }
}
```